src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -3001,16 +3003,30 @@ HCIMPL3(VOID, JIT_Unbox_Nullable, void * destPtr, CORINFO_CLASS_HANDLE type, Obj
 HCIMPLEND
 
 /*************************************************************/
-/* framed helper that handles full-blown type equivalence */
-NOINLINE HCIMPL2(LPVOID, JIT_Unbox_Helper_Framed, CORINFO_CLASS_HANDLE type, Object* obj)
+/* framed Unbox helper that handles enums and full-blown type equivalence */
+NOINLINE HCIMPL2(LPVOID, Unbox_Helper, CORINFO_CLASS_HANDLE type, Object* obj)
 {
     FCALL_CONTRACT;
 
     LPVOID result = NULL;
 
+    TypeHandle typeHnd(type);
+    // boxable types have method tables
+    _ASSERTE(!typeHnd.IsTypeDesc());
+    MethodTable* pMT1 = typeHnd.AsMethodTable();
+    MethodTable* pMT2 = obj->GetMethodTable();
+
     OBJECTREF objRef = ObjectToOBJECTREF(obj);
     HELPER_METHOD_FRAME_BEGIN_RET_1(objRef);
-    if (TypeHandle(type).IsEquivalentTo(objRef->GetTypeHandle()))
+
+    if (pMT1->GetInternalCorElementType() == pMT2->GetInternalCorElementType() &&
```

Comment:
You should be also able to make the enum case faster by using `GetVerifierCorElementType`. It should make the IsEnum/IsTruePrimitiveType conditions unnecessary.